### PR TITLE
Take the failed signature into consideration add for loop

### PR DESCRIPTION
**Pull request by xiaomingloveqinger**

### CODE REVIEW

File: blockchain/txvalidator.go [hunk 1]
```diff
@@ -1119,6 +1120,12 @@ func checkTransactionPayload(txn *Transaction) error {
 	case *payload.CoinBase:
 	case *payload.SideChainPow:
 	case *payload.WithdrawFromSideChain:
+		if txn.PayloadVersion == payload.WithdrawFromSideChainVersionV2 {
```
Comment (RainFallsSilent) [inline blockchain/txvalidator.go:1123]: need to move the check from sanity to context

Comment (xiaomingloveqinger) [reply blockchain/txvalidator.go]: Fixed it

**Review by Houshoupei84 — APPROVED** [review]

**Review by peterwillcn — APPROVED** [review]

File: crypto/schnorr.go [hunk 2]
```diff
@@ -184,13 +185,21 @@ func AggregateSignatures(privateKeys []*big.Int, message []byte) ([64]byte, erro
...
+func randomBytes(len int) []byte {
+	a := make([]byte, len)
+	rand.Read(a)
+	return a
+}
 
-	return k0, nil
+func deterministicGetK0(d []byte) (*big.Int, error) {
+	for {
+		message := randomBytes(32)
```
Comment (peterwillcn) [inline crypto/schnorr.go:196]: There's no need to introduce the function, merge it
```
message := make([]byte, 32)
rand.Read(message)
```

File: crypto/schnorr.go [hunk 3]
```diff
@@ -184,13 +185,21 @@ func AggregateSignatures(privateKeys []*big.Int, message []byte) ([64]byte, erro
...
 }
 
-func deterministicGetK0(d []byte, message []byte) (*big.Int, error) {
-	h := sha256.Sum256(append(d, message[:]...))
-	i := new(big.Int).SetBytes(h[:])
-	k0 := i.Mod(i, N)
-	if k0.Sign() == 0 {
-		return nil, errors.New("k0 is zero")
-	}
+func randomBytes(len int) []byte {
```
Comment (peterwillcn) [inline crypto/schnorr.go:188]: This isolated function, perhaps better incorporated into the reference.

**Review by peterwillcn — CHANGES_REQUESTED** [review]

**Review by peterwillcn — APPROVED** [review]

**Review by RainFallsSilent — APPROVED** [review]

**Pull request merged**